### PR TITLE
SPIKE - using a state machine for draft/live form

### PR DESCRIPTION
**Pull request by aliuk2012**
### States

```mermaid
stateDiagram
    Draft --> Live
    draft_live: Draft live
    Live --> draft_live
    draft_live --> Live
    draft_archived: Draft archived
    draft_live --> draft_archived
    Live --> Archived
    Archived --> Live
    Archived --> draft_archived
    draft_archived --> Live
 ```

#### Draft
Initial forms can be deleted as they are not official forms yet. Draft forms can be created/edited by trial/editor users. Previews of draft forms are viewable in `forms-runner`. This is the only state that a form created by a trial user can get to until the  trial users account has been upgraded to an Editor role.

#### Live
Form record is transitioned to this state when a user has completed all the required tasks for a form and decides to make a form live. Live forms cannot be HARD deleted as they are now official forms which could have been used by the public to submit their data to a government department. Previews of live forms are viewable in `forms-runner`.  When a form is made live we take a snapshot of the form, all its pages & all the page routing conditions and store it in `made_live_forms` table as JSON blob format. A form can have multiple "live versions" in `made_live_forms` table but we only ever return the latest version to our users. We keep multiple versions so we could serve up previous versions if need be or if users start a form and a new version is published, we should still be able to allow them to complete the version of form they started.

#### Draft live
Form record is transitioned to this state when a user clicks 
image
from the live form show page. This would be a change from the current method which is that we look at the forms `updated_at` and check if it is after the forms `live_at`.
A new draft of a live form but doesn't create a new record in the database. This state identifies when a form creator has made changes to a live form without affecting that live form. Once the form creator has approval for the changes to go live, the form can transition back to `Live` state.  Previews of draft forms are viewable in `forms-runner`. 

In this state, if a form is `archived`, the transition will go to `archived` and then on to `draft_archived` 

#### Archived
A form creator can only archive a form that is in either Live or Draft Live state. In this state a form filler would receive 404 "Page not Found" Error in forms-runner.  

#### Draft Archived
This is similar to how `Draft live` works except when  in this state a form filler would still receive 404 "Page not Found" Error in forms-runner unless they are previewing a draft version of the form. Once a form creator is ready to make a form live they should be able to transition this back to `Live` state.

[Trello](https://trello.com/c/G0bbEVGK/1447-spike-how-should-we-accurately-model-and-track-the-state-of-a-form)

- Adds a new column to forms table
- Adds aasm gem and after_commit_everywhere gem for activerecord support
- Mock out  a state machine for forms
	- reuse existing guards 		

Trello card: 

### Things to consider when reviewing

- Ensure that you consider the wider context.
- Does it work when run on your machine?
- Is it clear what the code is doing?
- Do the commit messages explain why the changes were made?
- Are there all the unit tests needed?
- Has all relevant documentation been updated?

### CODE REVIEW

File: app/state_machines/form_state_machine.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+module FormStateMachine
+  extend ActiveSupport::Concern
+
+  included do
+    include AASM
+
+    enum :state, {
+      draft: "draft",
+      live: "live",
```
Comment (aliuk2012) [inline app/state_machines/form_state_machine.rb:9]:
```suggestion
      initial_draft: "initial_draft",
      live: "live",
      draft: "draft"
```

I was wondering if would be be better than just draft/live because we can track how many forms are yet to be made live at all.

```
Form.initial_draft # returns all Forms that have never been made live

Form.live # returns all Forms that are made live

Form.draft # returns Forms that have changes since the form went live and are waiting for the changes to be made live
```

File: app/state_machines/form_state_machine.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    aasm column: :state, enum: true do
+      state :draft, initial: true
+      state :live
+
+      event :make_form_live do
+        transitions from: :draft, to: :live, guard: Proc.new { task_status_service.mandatory_tasks_completed? }
+      end
+
+      event :make_changes_live do
+        transitions from: :live, to: :draft, guard: Proc.new { has_live_version }
```
Comment (aliuk2012) [inline app/state_machines/form_state_machine.rb:21]: we do have business rules split across forms-admin and forms-api. For example in forms-admin, if live forms submission email changes when changes are made live then we notify the original submission email address.

File: Gemfile [hunk 3]
```diff
@@ -83,3 +83,7 @@ group :test do
 end
 
 gem "reverse_markdown", "~> 2.1"
+
+# Add state machine for forms
+gem "aasm", "~> 5.5"
+gem "after_commit_everywhere", "~> 1.0"
```
Comment (chao-xian) [inline Gemfile:89]: Why is this needed?

Comment (aliuk2012) [reply Gemfile:89]: The `after_commit_everywhere` gem is needed as mentioned in aasm readme https://github.com/aasm/aasm#activerecord. It is used to enabling persisting the the objects state when you call the event name as method with a bang. If the bang is not used then the object state is changed but is not saved in the DB

File: app/state_machines/form_state_machine.rb (hunk 2 above)
Comment (danworth) [inline app/state_machines/form_state_machine.rb:21]: I'd have expected `:make_changes_live` to transition `from: :draft, to: :live` but its currently the other way around?

Comment (aliuk2012) [reply app/state_machines/form_state_machine.rb:21]: I was trying to reuse the same language we have in the ui but yes you are correct in this instance it should actually be "draft" to "live" and there should be another event that gets triggered when anything on the form/pages/routing conditions changes.

I think having clear language between the states is what we need.

Another way is to think of "Live" forms as just "versions" of a Form. the latest version of the finalised form are published on govuk. the versioning is just a means for us to "make a commit"  of a form in its current state. This way our "Form" model works and "MadeLiveVersions" are snapshots of the form.

File: app/state_machines/form_state_machine.rb (hunk 1 above)
Comment (chao-xian) [reply app/state_machines/form_state_machine.rb:9]: * new
* live
* draft
?

File: app/state_machines/form_state_machine.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      draft: "draft",
+      live: "live",
+    }
+
+    aasm column: :state, enum: true do
+      state :draft, initial: true
+      state :live
+
+      event :make_form_live do
+        transitions from: :draft, to: :live, guard: Proc.new { task_status_service.mandatory_tasks_completed? }
```
[the remark this answers is not in the document]
Comment (chao-xian) [reply app/state_machines/form_state_machine.rb:17]: Nice!

File: app/state_machines/form_state_machine.rb (hunk 2 above)
Comment (aliuk2012) [reply app/state_machines/form_state_machine.rb]: resolving this conversation because the event/state names have changed a lot since this conversation took place.

File: app/models/condition.rb [hunk 5]
```diff
@@ -8,7 +8,10 @@ class Condition < ApplicationRecord
   has_one :form, through: :routing_page
 
   def save_and_update_form
-    save && form.update!(question_section_completed: false)
+    save
+    form.update!(question_section_completed: false)
+    form.draft_new_live_form! if form.live?
+    form.create_draft_from_archived_form! if form.archived?
```
Comment (aliuk2012) [inline app/models/condition.rb:14]: not going to need these changes if user triggers the state change themselves and specifically "creates a new draft"

File: app/models/form.rb [hunk 6]
```diff
@@ -41,7 +44,7 @@ def has_live_version
   end
 
   def live_version
-    raise ActiveRecord::RecordNotFound if made_live_forms.blank?
+    raise ActiveRecord::RecordNotFound if made_live_forms.blank? || archived?
```
Comment (aliuk2012) [inline app/models/form.rb:47]: may have to put in some other condition to check if runner or admin is trying to view last live version

File: app/state_machines/form_state_machine.rb (hunk 1 above)
Comment (aliuk2012) [reply app/state_machines/form_state_machine.rb:9]: resolving this as the state names have changed

File: app/state_machines/form_state_machine.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      draft: "draft",
+      live: "live",
+      draft_live: "draft_live",
+      archived: "archived",
+      draft_archived: "draft_archived",
+    }
+
+    aasm column: :state, enum: true do
+      state :draft, initial: true
+      state :live, :draft_live, :archived, :draft_archived
```
Comment (DavidBiddle) [inline app/state_machines/form_state_machine.rb:17]: Am I right in thinking these are roughly what the different states actually are? 
draft: Has no live snapshot
live: Matches most recent live snapshot
draft_live: Doesn't match most recent live snapshot
archived: Archived and matches live snapshot
draft_archived: Archived and doesn't match live snapshot

Comment (aliuk2012) [reply app/state_machines/form_state_machine.rb:17]: these are the values that would be stored in the `state` column of the forms table. A form would only ever have 1 state at a time. its not tied to any of the snapshots but you could still use `form.live_version` to get the snapshot (if it exists) in any states where you mentioned `snapshot` above

Comment (DavidBiddle) [reply app/state_machines/form_state_machine.rb:17]: Yeah that makes sense - was just checking whether I roughly understand what those states are meant to mean to us

**Comment by aliuk2012** [comment]
if you testing this locally in a rails console

you can check the state/events for a forms using this little lne

```ruby
f = Form.first #returns form record
{ states: f.aasm.states(permitted:true).map(&:name), events: f.aasm.events(permitted:true).map(&:name), state: f.aasm.current_state} #{:states=>[:archived], :events=>[:archive_live_form], :state=>:live}

```

File: app/state_machines/form_state_machine.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  included do
+    include AASM
+
+    enum :state, {
+      draft: "draft",
+      live: "live",
+      draft_live: "draft_live",
+      archived: "archived",
+      draft_archived: "draft_archived",
+    }
```
[the remark this answers is not in the document]
Comment (aliuk2012) [reply app/state_machines/form_state_machine.rb]: we can but we were trying to keep it as close to what users see when they view list of statuses

what do others thing about @DavidBiddle  suggestion? @alphagov/forms-devs

[the remark this answers is not in the document]
Comment (lfdebrux) [reply app/state_machines/form_state_machine.rb]: I also had comments about the transition names, but I put them on the other PR 😅 

https://github.com/alphagov/forms-api/pull/422#discussion_r1466117774

[the remark this answers is not in the document]
Comment (christophercameron-ixd) [reply app/state_machines/form_state_machine.rb]: Not a developer, but this seems like a good idea. It reads more explicit to me. 
And in terms of what the user sees this is basically how they will see it. They don’t know that it’s a single state - they’ll see it as 2 separate things, a draft AND a live/archived form

[the remark this answers is not in the document]
Comment (aliuk2012) [reply app/state_machines/form_state_machine.rb]: @thomasiles @jamie-o-wilkinson  and myself had a chat about this is and  were concerned about about how form-admin might map `live_with_draft` state to two different states for the users but it turns up this shouldn't be a problem because form-api forms model `has_draft_version` and `has_live_version` method which forms-admin uses to output the two different tags. we might want to change this in future but for the time being we happy to go with `live_with_draft` state. see implementation in https://github.com/alphagov/forms-api/pull/422

[the remark this answers is not in the document]
Comment (lfdebrux) [reply app/state_machines/form_state_machine.rb]:
```mermaid
stateDiagram-v2

    [*] --> Form

    state Form {
        [*] --> Draft
        Draft --> [*]
        --
        Archived --> Live
        Live --> Archived
    }

    Draft --> Live
```

File: app/state_machines/form_state_machine.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  included do
+    include AASM
+
+    enum :state, {
+      draft: "draft",
+      deleted: "deleted",
+      live: "live",
+      draft_live: "draft_live",
+      archived: "archived",
+      draft_archived: "draft_archived",
```
Comment (aliuk2012) [inline app/state_machines/form_state_machine.rb:13]:
```suggestion
      archived: "archived",
      archived_with_draft: "archived_with_draft",
```

File: app/state_machines/form_state_machine.rb [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    aasm column: :state, enum: true do
+      state :draft, initial: true
+      state :deleted, :live, :draft_live, :archived, :draft_archived
```
Comment (aliuk2012) [inline app/state_machines/form_state_machine.rb:18]:
```suggestion
      state :deleted, :live, :draft_live, :archived, :archived_with_draft
```
